File: QingToolbox.Android/app/src/main/java/com/qingtoolbox/android/QingTransferMetadata.kt
package com.qingtoolbox.android

object QingTransferMetadata {
    const val serviceType = "_qingtransfer._tcp"
    const val androidServiceType = "_qingtransfer._tcp."
    private const val protocolVersion = "1"
    private const val maxServiceNameLength = 255
    private const val maxFieldLength = 128
    private const val maxDisplayNameLength = 64

    sealed interface ParseResult {
        data class Valid(val peer: QingTransferPeer) : ParseResult
        data object Invalid : ParseResult
    }

    fun create(platform: String, displayName: String): Map<String, String> = mapOf(
        "v" to protocolVersion,
        "pf" to platform,
        "name" to sanitizeName(displayName),
        "cap" to "file",
    )

    fun parse(
        serviceName: String?,
        fields: Map<String, String?>,
        addresses: List<String> = emptyList(),
        port: Int,
        lastSeen: Long? = null,
    ): ParseResult {
        val normalized = serviceName?.trim()?.trimEnd('.') ?: return ParseResult.Invalid
        if (normalized.length > maxServiceNameLength ||
            !normalized.endsWith(".$serviceType.local", ignoreCase = true)
        ) return ParseResult.Invalid
        val version = field(fields, "v") ?: return ParseResult.Invalid
        val platform = field(fields, "pf") ?: return ParseResult.Invalid
        val name = field(fields, "name") ?: return ParseResult.Invalid
        val capabilities = field(fields, "cap")?.split(',')
            ?.map(String::trim)
            ?.filter { it.isNotEmpty() && it.length <= maxFieldLength }
            ?.distinctBy(String::lowercase)
            ?: return ParseResult.Invalid
        if (version != protocolVersion || platform !in setOf("windows", "android") ||
            name.length !in 1..maxDisplayNameLength || name.any(Char::isISOControl) ||
            capabilities.none { it.equals("file", ignoreCase = true) } ||
            port !in 0..65535
        ) return ParseResult.Invalid
        return ParseResult.Valid(
            QingTransferPeer(
                serviceName = normalized,
                displayName = name,
                platform = platform,
                protocolVersion = version,
                capabilities = capabilities,
                addresses = addresses.asSequence().filter(::isSafeAddress).distinct().take(8).toList(),
                port = port,
                online = true,
                lastSeen = lastSeen,
            ),
        )
    }

    fun fullServiceName(name: String?): String {
        val value = name?.trim()?.trimEnd('.') ?: return ""
        return when {
            value.endsWith(".$serviceType.local", ignoreCase = true) -> value
            value.endsWith(".$serviceType", ignoreCase = true) -> "$value.local"
            else -> "$value.$serviceType.local"
        }
    }

    fun sanitizeName(value: String): String {
        val safe = value.trim().filter { it.isLetterOrDigit() || it == '-' || it == '_' }
        return (if (safe.isEmpty()) "QingToolbox" else safe).take(maxDisplayNameLength)
    }

    private fun field(fields: Map<String, String?>, key: String): String? {
        val value = fields.entries.firstOrNull { it.key.equals(key, ignoreCase = true) }?.value?.trim() ?: return null
        return value.takeIf { it.isNotEmpty() && it.length <= maxFieldLength && it.none(Char::isISOControl) }
    }

    private fun isSafeAddress(value: String): Boolean =
        value.length <= 64 && value.isNotBlank() && value.none(Char::isISOControl)
}

File: modules/QingTransfer/QingTransferMetadata.cs
using System.Collections.ObjectModel;
using System.Net;

namespace QingToolbox.Modules.QingTransfer;

internal static class QingTransferMetadata
{
    internal const string ServiceType = "_qingtransfer._tcp";
    internal const string WindowsServiceType = "_qingtransfer._tcp.local";
    internal const string AndroidServiceType = "_qingtransfer._tcp.";
    internal const string ProtocolVersion = "1";
    private const int MaxServiceNameLength = 255;
    private const int MaxFieldLength = 128;
    private const int MaxDisplayNameLength = 64;

    public static IReadOnlyDictionary<string, string> Create(string platform, string displayName) =>
        new ReadOnlyDictionary<string, string>(new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
        {
            ["v"] = ProtocolVersion,
            ["pf"] = platform,
            ["name"] = displayName,
            ["cap"] = "file",
        });

    public static QingTransferPeer? Parse(
        string? serviceName,
        IReadOnlyDictionary<string, string?> fields,
        string? hostName,
        int port,
        IEnumerable<IPAddress>? addresses = null,
        DateTimeOffset? lastSeen = null)
    {
        if (string.IsNullOrWhiteSpace(serviceName)) return null;
        var normalizedServiceName = serviceName.Trim().TrimEnd('.');
        if (normalizedServiceName.Length > MaxServiceNameLength) return null;
        if (!normalizedServiceName.EndsWith("._qingtransfer._tcp.local", StringComparison.OrdinalIgnoreCase)) return null;
        if (!TryField(fields, "v", out var version) || version != ProtocolVersion) return null;
        if (!TryField(fields, "pf", out var platform) ||
            (platform != "windows" && platform != "android")) return null;
        if (!TryField(fields, "name", out var name) || !IsSafeDisplayName(name)) return null;
        if (!TryField(fields, "cap", out var capabilityText)) return null;
        var capabilities = capabilityText
            .Split(',', StringSplitOptions.TrimEntries | StringSplitOptions.RemoveEmptyEntries)
            .Where(value => value.Length <= MaxFieldLength)
            .Distinct(StringComparer.OrdinalIgnoreCase)
            .ToArray();
        if (!capabilities.Contains("file", StringComparer.OrdinalIgnoreCase)) return null;
        if (port is < 0 or > ushort.MaxValue) return null;

        var safeAddresses = (addresses ?? []).Distinct().Take(8).ToArray();
        return new QingTransferPeer(
            normalizedServiceName,
            name,
            platform,
            version,
            capabilities,
            safeAddresses,
            port,
            Online: true,
            lastSeen);
    }

    private static bool TryField(
        IReadOnlyDictionary<string, string?> fields,
        string key,
        out string value)
    {
        value = string.Empty;
        if (!fields.TryGetValue(key, out var candidate) || string.IsNullOrWhiteSpace(candidate)) return false;
        candidate = candidate.Trim();
        if (candidate.Length > MaxFieldLength || candidate.Any(char.IsControl)) return false;
        value = candidate;
        return true;
    }

    private static bool IsSafeDisplayName(string value) =>
        value.Length is > 0 and <= MaxDisplayNameLength && !value.Any(char.IsControl);
}
